Parametrize break sequence for word_wrap on ActionView Text Helpers

bukue opened the pull request:
Sometimes you need a specific break sequence while using word wrap and as today the only option we have is "\n" and is hardcoded.

With this change you will be able to specify any break sequence ("\r\n" for example) as an option.

meinac commented:
@bukue could you please add documentation about it?

bukue commented:
@meinac done!

kaspth commented on actionview/test/template/text_helper_test.rb:
```diff
@@ -366,6 +366,10 @@ def test_word_wrap_does_not_modify_the_options_hash
   end
 
+  def test_word_wrap_with_custom_break_sequence
+    assert_equal("1234567890\r\n1234567890\r\n1234567890", word_wrap("1234567890 "*3, :line_width => 2, :break_sequence => "\r\n"))
```
New hash syntax please. And add spaces around `*`.

bukue commented:
@meinac @kaspth I just submitted the changes that you guys asked.  This is ready for another review.

kaspth commented:
@bukue doing great. Now we just need you to squash your commits down to one :grin:

bukue commented:
@kaspth done!

matthewd commented:
I'm a bit dubious as to whether we actually want this... but that notwithstanding, shouldn't the `split` change too?

bukue commented:
@matthewd Why are you dubious? 

The main reason to split with "\n" is to split by lines. I see how it would be useful to split it by "\r\n" or any other sequence that helps defining a line. Unless somebody else considers it not useful, I will gladly implement it.

bukue commented:
@matthewd added the changes you suggested

matthewd commented:
Why would your `line_splitter` and `break_sequence` differ? My concern was that you'd seemingly-arbitrarily chosen to replace *some* of the instances of `"\n"` in the method with a parameter -- not that we hadn't added enough parameters.

What does this offer over `word_wrap(...).gsub("\n", "\r\n")` ?

If I use a `break_sequence` of `"...\nand then, "`, how does that interact with the line width? I can obviously see the answer... but is that going to be what callers would expect, and/or find most useful?

"Great!, this is just what I need... I'll wrap on `"<br>\n".html_safe`"

bukue commented:
Yeah, you can totally have the same result doing word_wrap(...).gsub("\n", "\r\n"). Actually, this PR started because we were doing the exact same thing at work because we needed the end of lines to be indicated with "\r\n".

I see your point, although I feel this changes give you more legibility and flexibility. Let's see what the other guys think.

bukue commented:
@kaspth @meinac what do you guys think?

kaspth commented:
This feels like way too much code for me for something that can be done with `gsub` ¯\_(ツ)_/¯

meinac commented:
I agree with @kaspth. If you are doing this many times in your application then I recommend you to create a method which does what do you want for your application.

bukue commented:
This approach is cheaper since you don't have to transverse your string twice like you would do using gsub. Also, all I am doing is parametrizing things that otherwise would be hardcoded.

This would be my first commit to the rails repository so I might be going to far with my changes. So if you guys believe this change is too much I am happy to roll it back to just parametrizing the line break. 

What do you guys think?

kaspth commented on actionview/lib/action_view/helpers/text_helper.rb:
```diff
@@ -250,12 +250,24 @@ def pluralize(count, singular, plural = nil, locale: nil)
+      #   # => "Once upon\na time,\nin a\nkingdom\ncalled Far\nFar Away,\na king\nfell\nill..."
       def word_wrap(text, options = {})
         line_width = options.fetch(:line_width, 80)
+        line_splitter = options.fetch(:line_splitter, "\n")
```
Lets take out the line splitter.

kaspth commented on actionview/lib/action_view/helpers/text_helper.rb:
```diff
@@ -250,12 +250,24 @@ def pluralize(count, singular, plural = nil, locale: nil)
+      #
+      #   word_wrap('Once upon a time,| in a kingdom called Far Far Away,| a king fell ill...', line_width: 10, line_splitter: "|")
+      #   # => "Once upon\na time,\nin a\nkingdom\ncalled Far\nFar Away,\na king\nfell\nill..."
       def word_wrap(text, options = {})
```
When we're changing existing code it's okay to convert to keyword arguments like this:

```ruby
def word_wrap(text, line_with: 80, break_sequence: "\n")
```

Commit:
Sometimes you need a specific break sequence while using word wrap and as today the only option we have is "\n" and is hardcoded.

With this change you will be able to specify any break sequence ("\r\n" for example) as an option.

adding proper documentation for break_sequence in ActionView::Helpers::TextHelper.word_wrap

adding some more documentation for word_wrap custom break sequence and making sure we use new hash syntax

bukue commented:
@kaspth All done!

kaspth commented:
Thanks
